Libplanet/Action/ActionEvaluator.cs

```diff
@@ -0,0 +1,581 @@
+        /// being executed.  Note that <see cref="IAccountStateDelta.UpdatedAddresses"/> of
+        /// <paramref name="previousStates"/> are remained to the returned next states.</param>
+        /// <param name="miner">An address of block miner.</param>
+        /// <param name="signer">Signer of the <paramref name="actions"/>.</param>
+        /// <param name="signature"><see cref="Transaction{T}"/> signature used to generate random
+        /// seeds.</param>
+        /// <param name="actions">Actions to evaluate.</param>
+        /// <param name="rehearsal">Pass <c>true</c> if it is intended
+        /// to be dry-run (i.e., the returned result will be never used).
+        /// The default value is <c>false</c>.</param>
+        /// <param name="previousBlockStatesTrie">The trie to contain states at previous block.
+        /// </param>
+        /// <param name="blockAction">Pass <c>true</c> if it is
+        /// <see cref="IBlockPolicy{T}.BlockAction"/>.</param>
+        /// <returns>Enumerates <see cref="ActionEvaluation"/>s for each one in
+        /// <paramref name="actions"/>.  The order is the same to the <paramref name="actions"/>.
+        /// Note that each <see cref="IActionContext.Random"/> object
+        /// has a unconsumed state.
+        /// </returns>
+        [Pure]
+        internal static IEnumerable<ActionEvaluation> EvaluateActions(
+            BlockHash preEvaluationHash,
+            long blockIndex,
+            TxId? txid,
+            IAccountStateDelta previousStates,
+            Address miner,
+            Address signer,
+            byte[] signature,
+            IImmutableList<IAction> actions,
+            bool rehearsal = false,
+            ITrie? previousBlockStatesTrie = null,
+            bool blockAction = false)
+        {
+            ActionContext CreateActionContext(IAccountStateDelta prevStates, int randomSeed)
+            {
+                return new ActionContext(
+                    signer: signer,
+                    txid: txid,
+                    miner: miner,
+                    blockIndex: blockIndex,
+                    previousStates: prevStates,
+                    randomSeed: randomSeed,
+                    rehearsal: rehearsal,
+                    previousBlockStatesTrie: previousBlockStatesTrie,
+                    blockAction: blockAction);
+            }
+
+            byte[] hashedSignature;
+            using (var hasher = SHA1.Create())
+            {
+                hashedSignature = hasher.ComputeHash(signature);
+            }
+
+            byte[] preEvaluationHashBytes = preEvaluationHash.ToByteArray();
+            int seed =
+                (preEvaluationHashBytes.Length > 0
+                    ? BitConverter.ToInt32(preEvaluationHashBytes, 0) : 0)
+                ^ (signature.Any() ? BitConverter.ToInt32(hashedSignature, 0) : 0);
+
+            IAccountStateDelta states = previousStates;
+            ILogger logger = Log.ForContext<ActionEvaluation>();
```

Comment:
Why not use `_logger` instead?

Reply:
Partly because `EvaluateActions()` is left as `static`. Not sure if it was the best design choice, but the my reasoning behind it was to leave last mile/heavy lifting as `static` since this may be considered more of as a utility method, and possibly for optimization.

Reply:
As logger contexts are for filtering logs, to me the best practice is to make logs belong to the context where the logging code is placed file-wise.  The current context implies the logging code is placed in *ActionEvaluation.cs*, it would be rather a trap when we investigate issues in production.

Reply:
Well, I'll just turn the method into an instance method. Shouldn't affect the performance too hard anyways.

I think a more coherent solution would be to make it `private` and handle it from outside.

Reply:
Eh... ended up making `_logger` static instead. See 87849869dbdab8c8548cc56db3e054bd1dd291d9.